### PR TITLE
Fix visualization for semantic segmentation and add confidence thresholding

### PR DESCRIPTION
**Pull request by manushreegangwar**
## What changes are proposed in this pull request?

- Enforce class indices to start at 1 for models where class index 0 doesn't correspond to background class. This is to fix semantic segmentation mask visualization (app doesn't visualize mask pixels set to 0 class index).
- Enable confidence thresholding for semantic segmentation masks so that any mask pixel with confidence below threshold is assigned to class index 0.

## How is this patch tested? If it is not, please explain why.

```
# Semantic segmentation models in the zoo
'deeplabv3-resnet101-coco-torch',
 'deeplabv3-resnet50-coco-torch',
 'fcn-resnet101-coco-torch',
 'fcn-resnet50-coco-torch',
 'group-vit-segmentation-transformer-torch',
 'segformer-b0-ade20k-torch',
 'segformer-b1-ade20k-torch',
 'segformer-b2-ade20k-torch',
 'segformer-b3-ade20k-torch',
 'segformer-b4-ade20k-torch',
 'segformer-b5-ade20k-torch',
 'segmentation-transformer-torch'
```

1. Run semantic segmentation for all models with multiple confidence thresolds
```
import fiftyone as fo
import fiftyone.operators as foo
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart")

for model_name in seg_models:
    if model_name == 'group-vit-segmentation-transformer-torch':
        model = foz.load_zoo_model("group-vit-segmentation-transformer-torch",
                    text_prompt="a photo of a",
                    classes=["cat", "dog", "bear", "person", "vehicle", "other"])
    else:
        model = foz.load_zoo_model(model_name)

    for conf in [0.0, 0.25, 0.8]:
        label_field = (f"{model_name}_conf{conf}").replace(".","")
        print(label_field)
        dataset.mask_targets[label_field] = model.mask_targets
        dataset.apply_model(model, label_field=label_field, batch_size=8, confidence_thresh=conf)
```

2. Run semantic segmentation on `develop` and compare output against this branch for confidence threshold set to 0.0
```
for model_name in seg_models[:-1]:
    if model_name == 'group-vit-segmentation-transformer-torch':
        model = foz.load_zoo_model("group-vit-segmentation-transformer-torch",
                    text_prompt="a photo of a",
                    classes=["cat", "dog", "bear", "person", "vehicle", "other"])
    else:
        model = foz.load_zoo_model(model_name)

    label_field = (f"{model_name}_dev")
    print(label_field)
    if model.mask_targets:
        dataset.mask_targets[label_field] = model.mask_targets
    dataset.apply_model(model, label_field=label_field, batch_size=8, skip_failures=False)

# For models with no 0 in mask target keys, increment masks by 1 for evaluation (groupvit, segmentation-transformer-torch)

for model_name in seg_models:
    label_field = (f"{model_name}_dev")
    br_label_field = (f"{model_name}_conf00")
    results = dataset.evaluate_segmentations(
        br_label_field,
        gt_field=label_field,
        eval_key=model_name.replace("-","_"),
    )
    print(results.print_metrics())
```

3. Qualitative evaluation in the app to check whether mask visualization looks good.

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

All semantic segmentation torch models support confidence thresholding and the class indices start at 1 for segmentation masks.
 
### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [x] Other

## Summary by CodeRabbit

* **New Features**
  * Option to reserve a dedicated background index in semantic segmentation outputs.
  * Per-pixel confidence thresholding to mark low-confidence pixels as background.
  * Accepts either raw logits or softmaxed outputs and supports more flexible input arguments.

* **Improvements**
  * Mask-to-class mappings now derive from model classes and respect background reservation.

### CODE REVIEW

File: fiftyone/utils/torch.py [hunk 1]
```diff
@@ -1623,8 +1638,21 @@ def __call__(self, output, *args, **kwargs):
         Returns:
             a list of :class:`fiftyone.core.labels.Segmentation` instances
         """
-        probs = output["out"].detach().cpu().numpy()
+        out = output["out"].detach().cpu()
+        if not (torch.all(out >= 0.0) and torch.all(out <= 1.0)):
```
Comment (PeterOttoMoeller) [inline fiftyone/utils/torch.py]: Is this because some models already apply softmax? Would it be less brittle/more straightforward to add some kind of configuration parameter for these two cases (softmax already applied or not) instead of this numeric check?

File: fiftyone/utils/torch.py [hunk 2]
```diff
@@ -1623,8 +1638,21 @@ def __call__(self, output, *args, **kwargs):
         Returns:
             a list of :class:`fiftyone.core.labels.Segmentation` instances
         """
-        probs = output["out"].detach().cpu().numpy()
+        out = output["out"].detach().cpu()
```
Comment (PeterOttoMoeller) [inline fiftyone/utils/torch.py:1649]: Could we keep all the following tensor operations that were added with this PR on the GPU? I am not sure if that would actually have a runtime advantage, but they all seem to be quite parallelizable

File: fiftyone/utils/torch.py [hunk 3]
```diff
@@ -963,6 +963,19 @@ def _parse_mask_targets(self, config):
             mask_targets_path = fou.fill_patterns(config.mask_targets_path)
             return etal.load_labels_map(mask_targets_path)
 
+        if self.classes and config.output_processor_args:
+            if config.output_processor_args.get("no_background_cls", False):
```
Comment (PeterOttoMoeller) [inline fiftyone/utils/torch.py:971]: If the model DOES output a background class, we set `mask_targets` to start at 1 for the first class in `classes`. Do we know for certain that `classes` does not include "backgorund" as a class? 

In general, does the order of classes in `classes` always exactly correspond to the model's output? (Assuming this is indeed enforced somewhere else or simply has to be assumed to be true through correct configuration -- just wanting to make sure)

File: fiftyone/zoo/models/manifest-torch.json [hunk 4]
```diff
@@ -2769,7 +2769,11 @@
             "size_bytes": null,
             "default_deployment_config_dict": {
                 "type": "fiftyone.utils.transformers.FiftyOneTransformerForSemanticSegmentation",
-                "config": {}
+                "config": {
+                    "output_processor_args": {
+                        "no_background_cls": true
```
Comment (PeterOttoMoeller) [inline fiftyone/zoo/models/manifest-torch.json]: Potentially dumb/obvious question: how did you find out that only these 2 specific models `segmentation-transformer-torch` and `group-vit-segmentation-transformer-torch` do not output 0 as background?

**Comment by PeterOttoMoeller** [comment]
I am running this 

`semseg_models = foz.list_zoo_models(tags="segmentation")
# Remove tensorflow models
semseg_models = [model for model in semseg_models if not model.endswith("-tf")]

print(semseg_models)`

on current develop and am getting more than your 6 models in the PR description:

['deeplabv3-resnet101-coco-torch', 'deeplabv3-resnet50-coco-torch', 'fcn-resnet101-coco-torch', 'fcn-resnet50-coco-torch', 'group-vit-segmentation-transformer-torch', 'segformer-b0-ade20k-torch', 'segformer-b1-ade20k-torch', 'segformer-b2-ade20k-torch', 'segformer-b3-ade20k-torch', 'segformer-b4-ade20k-torch', 'segformer-b5-ade20k-torch', 'segmentation-transformer-torch']

File: fiftyone/utils/torch.py (hunk 3 above)
Comment (manushreegangwar) [reply fiftyone/utils/torch.py:971]: Good question. This change doesn't take into consideration whether there is a "background" class in the model config or not. If the class at index "0" is background (in 'deeplabv3-resnet101-coco-torch' as an example), we will skip incrementing if `no_background_cls = False`. We assume that background class will be class index 0 or assumed to be not present. If there is "background"/ "other" class at any other index, it will be visualized. 

If there's a need to configure classes that are to be treated as class index 0, we can introduce a new parameter to configure such classes. I would do it as a separae PR. 

>  does the order of classes in classes always exactly correspond to the model's output?

https://github.com/voxel51/fiftyone/blob/develop/fiftyone/utils/torch.py#L952-L966
This is not strictly enforced but is assumed that the model configuration is done correctly. Also, `mask_targets` doesn't affect model inference. It is only used for [visualizing per-pixel labels in the app.](https://github.com/voxel51/fiftyone/blob/develop/fiftyone/utils/torch.py#L952-L966). Note that line 966 - 977 are not  required for visualizations to work correctly.

File: fiftyone/utils/torch.py (hunk 2 above)
Comment (manushreegangwar) [reply fiftyone/utils/torch.py:1649]: Good point. I will move softmax to gpu.

File: fiftyone/utils/torch.py (hunk 1 above)
Comment (manushreegangwar) [reply fiftyone/utils/torch.py]: Do you mean add it as a model configuration parameter? Yes, that works.

File: fiftyone/zoo/models/manifest-torch.json (hunk 4 above)
Comment (manushreegangwar) [reply fiftyone/zoo/models/manifest-torch.json]: If you look at the visualization in the app, you can see that class at index 0 (let's say "cat") is not visualized correctly. More details in jira about this.

File: fiftyone/utils/torch.py (hunk 3 above)
Comment (PeterOttoMoeller) [reply fiftyone/utils/torch.py:971]: Great, sounds good to me, feel free to resolve conversation

File: fiftyone/zoo/models/manifest-torch.json [hunk 5]
```diff
@@ -89,7 +92,14 @@
...
             },
-            "tags": ["segmentation", "coco", "torch", "resnet", "deeplabv3", "official"],
+            "tags": [
+                "segmentation",
+                "coco",
+                "torch",
+                "resnet",
+                "deeplabv3",
+                "official"
+            ],
```
Comment (manushreegangwar) [inline fiftyone/zoo/models/manifest-torch.json:102]: I can undo the pylint changes if needed but the pre-commit hook is applying these.

File: fiftyone/utils/torch.py (hunk 2 above)
Comment (manushreegangwar) [reply fiftyone/utils/torch.py:1649]: Done

File: fiftyone/utils/torch.py (hunk 1 above)
Comment (manushreegangwar) [reply fiftyone/utils/torch.py]: Added `has_softmax_out` to the output processor. All segmentation models currently in the zoo require this to be true.

File: fiftyone/utils/torch.py [hunk 6]
```diff
@@ -1623,8 +1642,21 @@ def __call__(self, output, *args, **kwargs):
...
             a list of :class:`fiftyone.core.labels.Segmentation` instances
         """
-        probs = output["out"].detach().cpu().numpy()
+        out = output["out"]
+        if not self.has_softmax_out:
+            out = out.softmax(dim=1)
+        probs = out.detach().cpu().numpy()
+
         masks = probs.argmax(axis=1)
+        if self.no_background_cls:
```
Comment (PeterOttoMoeller) [inline fiftyone/utils/torch.py:1655]: Could all the following tensor operations not also stay on the GPU? Essentially till the return

File: fiftyone/zoo/models/manifest-torch.json (hunk 5 above)
Comment (PeterOttoMoeller) [reply fiftyone/zoo/models/manifest-torch.json:102]: Seems more readable to me this way, would keep

File: fiftyone/utils/torch.py (hunk 6 above)
Comment (manushreegangwar) [reply fiftyone/utils/torch.py:1655]: They may but since these are not very heavy operations it is probably better to move them to CPU

Comment (manushreegangwar) [reply fiftyone/utils/torch.py:1655]: I have moved all post-processing to cpu for this PR but let's continue the discussion on GPU post-processing and tackle that in a separate PR.
